### PR TITLE
[python3 migration] Added Python3 support for new test cases

### PR DESCRIPTION
**Pull request by OleksandrKozodoi**
Signed-off-by: Oleksandr Kozodoi <oleksandrx.kozodoi@intel.com>

#### This PR includes necessary changes from [#9277](https://github.com/Azure/sonic-buildimage/pull/9277) for the setup of the Python3  virtual environment in the sonic-mgmt docker container.
Summary:
- Added Python3 support for lldp test case.
- Added Python3 support for dhcp_relay test case.
- Added Python3 support for decap test case.
- Added Python3 support for fib test case.
- Added Python3 support for bgp_bounce test case.
### How to use Python3 environment for running test cases?
1. Connect to the sonic-mgmt container
```
$ docker exec -ti sonic-mgmt bash
```
2. Activate the virtual environment
```
$ source /var/user/env-python3/bin/activate
```
3. Deploy t0/t1 topology
4. Run test case
```
py.test --testbed=testbed-t0 --inventory=../ansible/lab --testbed_file=../ansible/testbed.csv --host-pattern=testbed-t0 -- module-path=../ansible/library lldp
```
```
py.test --testbed=testbed-t0 --inventory=../ansible/lab --testbed_file=../ansible/testbed.csv --host-pattern=testbed-t0 -- module-path=../ansible/library dhcp_relay
```
```
py.test --testbed=testbed-t0 --inventory=../ansible/lab --testbed_file=../ansible/testbed.csv --host-pattern=testbed-t0 -- module-path=../ansible/library decap
```
```
py.test --testbed=testbed-t0 --inventory=../ansible/lab --testbed_file=../ansible/testbed.csv --host-pattern=testbed-t0 -- module-path=../ansible/library fib
```
```
py.test --testbed=testbed-t0 --inventory=../ansible/lab --testbed_file=../ansible/testbed.csv --host-pattern=testbed-t0 -- module-path=../ansible/library bgp/test_bgp_bounce.py
```
### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911

#### What is the motivation for this PR?
Migration of sonic-mgmt codebase from Python 2 to Python 3
#### How did you do it?
Updated codebase of sonic-mgmt for supporting Python3
#### How did you verify/test it?
Run test cases. Tests passed.
#### Any platform specific information?
SONiC Software Version: SONiC.master.60600-dirty-20211221.191419
Distribution: Debian 11.2
Kernel: 5.10.0-8-2-amd64
Build commit: 67e40b553
Build date: Tue Dec 21 19:22:34 UTC 2021
ASIC: barefoot
#### Supported testbed topology if it's a new test case?
t0, t1

### CODE REVIEW

File: ansible/library/lldp_facts.py [hunk 1]
```diff
@@ -258,7 +258,7 @@ def main():
 
     lldp_data = dict()
 
-    for intf in lldp_rem_sys.viewkeys():
+    for intf in lldp_rem_sys.keys():
```
Comment (novikauanton) [inline ansible/library/lldp_facts.py:261]: for loop will iter over keys by default. keys method call is redundant.

File: tests/common/devices/sonic.py [hunk 2]
```diff
@@ -1066,10 +1066,10 @@ def get_ip_route_info(self, dstip, ns=""):
 
             m = re.match(".+\s+via\s+(\S+)\s+.*dev\s+(\S+)\s+.*src\s+(\S+)\s+", rt[0])
             if m:
-                nexthop_ip = ipaddress.ip_address(unicode(m.group(1)))
+                nexthop_ip = ipaddress.ip_address(m.group(1).encode().decode("utf-8"))
```
Comment (novikauanton) [inline tests/common/devices/sonic.py:1069]: you can use six for porting code from py2 to py3 and wise verse. For example something like  [six.text_type](https://six.readthedocs.io/#six.text_type)

File: ansible/library/lldp_facts.py (hunk 1 above)
Comment (OleksandrKozodoi) [reply ansible/library/lldp_facts.py]: Done. Thanks

File: tests/common/devices/sonic.py (hunk 2 above)
Comment (OleksandrKozodoi) [reply tests/common/devices/sonic.py]: Done

**Comment by OleksandrKozodoi** [comment]
@yxieca @wangxin Can you take a look? Thank you in advance.

File: tests/common/errors.py [hunk 3]
```diff
@@ -26,8 +26,7 @@ def __init__(self, msg, results=None):
         self.results = results
 
     def _to_string(self):
-        return unicode(u"{}, Ansible Results =>\n{}".format(self.message, dump_ansible_results(self.results)))\
-            .encode('ascii', 'backslashreplace')
+        return "{}, Ansible Results =>\n{}".format(self.message, dump_ansible_results(self.results)).encode().decode("utf-8")
```
Comment (saiarcot895) [inline tests/common/errors.py:29]: What's the reason for calling `encode()` followed by `decode("utf-8")` here?

Comment (OleksandrKozodoi) [reply tests/common/errors.py:29]: This approach is used for backward compatibility of code between different versions of Python. Ansible returns different types of dump depending on Python versions.

**Comment by wangxin** [comment]
@OleksandrKozodoi Thanks for the efforts of making the scripts working for python3!
Have you tested the changes using python2? I believe there are more compatibilities issues in the repository. Before all the issues are resolved, we need to stick with python2 for some time. So, all the changes need to be both python2&3 compatible.

**Comment by OleksandrKozodoi** [comment]
@wangxin All scripts were tested using python2 and python3 versions. All necessary dependencies for testing python2&3 compatible will be available in python venv of the sonic-mgmt docker container after merging [9277](https://github.com/Azure/sonic-buildimage/pull/9277). These changes allow running tests on different versions in a single sonic-mgmt container. Thanks!

**Review by wangxin — DISMISSED** [review]

**Comment by wangxin** [comment]
LGTM, @OleksandrKozodoi Can you help resolve the conflicts?

**Comment by OleksandrKozodoi** [comment]
@wangxin I've resolved conflicts and added some minor changes for supporting the original codebase. But should be merged [9277](https://github.com/Azure/sonic-buildimage/pull/9277) PR of sonic-buildimage repo for checking compatibility of new changes with python3. Thanks.

**Review by ZhaohuiS — APPROVED** [review]

**Comment by ppikh** [comment]
Hi @OleksandrKozodoi 

Looks like this commit cause failures: TypeError: <lambda>() takes exactly 2 arguments (1 given)

Issue happen when run tests from folder: qos/test_qos_sai.py

It's related to file: tests/common/fixtures/ptfhost_utils.py

Example of python traceback:
```
duthosts = [<MultiAsicSonicHost arc-switch1038>], rand_one_dut_hostname = 'arc-switch1038', ptfhost = <tests.common.devices.ptf.PTFHost object at 0x7f60d0924590>

    @pytest.fixture(scope='class')
    def ptf_portmap_file(duthosts, rand_one_dut_hostname, ptfhost):
        """
            Prepare and copys port map file to PTF host
    
            Args:
                request (Fixture): pytest request object
                duthost (AnsibleHost): Device Under Test (DUT)
                ptfhost (AnsibleHost): Packet Test Framework (PTF)
    
            Returns:
                filename (str): returns the filename copied to PTF host
        """
        duthost = duthosts[rand_one_dut_hostname]
        intfInfo = duthost.show_interface(command = "status")['ansible_facts']['int_status']
        portList = natsorted([port for port in intfInfo if port.startswith('Ethernet')])
        portMapFile = "/tmp/default_interface_to_front_map.ini"
        with open(portMapFile, 'w') as file:
            file.write("# ptf host interface @ switch front port name\n")
            file.writelines(
                map(
                        lambda index, port: "{0}@{1}\n".format(index, port),
>                       enumerate(portList)
                    )
                )
E           TypeError: <lambda>() takes exactly 2 arguments (1 given)

duthost    = <MultiAsicSonicHost arc-switch1038>
duthosts   = [<MultiAsicSonicHost arc-switch1038>]
file       = <closed file '/tmp/default_interface_to_front_map.ini', mode 'w' at 0x7f60d028cd20>
intfInfo   = {'Ethernet0': {'admin_state': u'up', 'alias': u'etp1', 'fec': u'rs', 'name': u'Ethernet0', ...}, 'Ethernet100': {'admi...et104', ...}, 'Ethernet108': {'admin_state': u'up', 'alias': u'etp28', 'fec': u'rs', 'name': u'Ethernet108', ...}, ...}
port       = 'Ethernet68'
portList   = ['Ethernet0', 'Ethernet4', 'Ethernet8', 'Ethernet12', 'Ethernet16', 'Ethernet20', ...]
portMapFile = '/tmp/default_interface_to_front_map.ini'
ptfhost    = <tests.common.devices.ptf.PTFHost object at 0x7f60d0924590>
rand_one_dut_hostname = 'arc-switch1038'

common/fixtures/ptfhost_utils.py:186: TypeError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /root/mars/workspace/sonic-mgmt/tests/common/fixtures/ptfhost_utils.py(186)ptf_portmap_file()
-> enumerate(portList)
(Pdb) portList
[u'Ethernet0', u'Ethernet4', u'Ethernet8', u'Ethernet12', u'Ethernet16', u'Ethernet20', u'Ethernet24', u'Ethernet28', u'Ethernet32', u'Ethernet36', u'Ethernet40', u'Ethernet44', u'Ethernet48', u'Ethernet52', u'Ethernet56', u'Ethernet60', u'Ethernet64', u'Ethernet65', u'Ethernet66', u'Ethernet67', u'Ethernet68', u'Ethernet72', u'Ethernet76', u'Ethernet80', u'Ethernet82', u'Ethernet84', u'Ethernet88', u'Ethernet92', u'Ethernet96', u'Ethernet100', u'Ethernet104', u'Ethernet108', u'Ethernet112', u'Ethernet116', u'Ethernet120', u'Ethernet124']
(Pdb) index
*** NameError: name 'index' is not defined
(Pdb) port
u'Ethernet68'

```

**Comment by OleksandrKozodoi** [comment]
> Hi @OleksandrKozodoi
> 
> Looks like this commit cause failures: TypeError: () takes exactly 2 arguments (1 given)
> 
> Issue happen when run tests from folder: qos/test_qos_sai.py
> 
> It's related to file: tests/common/fixtures/ptfhost_utils.py
> 
> Example of python traceback:
> 
> ```
> duthosts = [<MultiAsicSonicHost arc-switch1038>], rand_one_dut_hostname = 'arc-switch1038', ptfhost = <tests.common.devices.ptf.PTFHost object at 0x7f60d0924590>
> 
>     @pytest.fixture(scope='class')
>     def ptf_portmap_file(duthosts, rand_one_dut_hostname, ptfhost):
>         """
>             Prepare and copys port map file to PTF host
>     
>             Args:
>                 request (Fixture): pytest request object
>                 duthost (AnsibleHost): Device Under Test (DUT)
>                 ptfhost (AnsibleHost): Packet Test Framework (PTF)
>     
>             Returns:
>                 filename (str): returns the filename copied to PTF host
>         """
>         duthost = duthosts[rand_one_dut_hostname]
>         intfInfo = duthost.show_interface(command = "status")['ansible_facts']['int_status']
>         portList = natsorted([port for port in intfInfo if port.startswith('Ethernet')])
>         portMapFile = "/tmp/default_interface_to_front_map.ini"
>         with open(portMapFile, 'w') as file:
>             file.write("# ptf host interface @ switch front port name\n")
>             file.writelines(
>                 map(
>                         lambda index, port: "{0}@{1}\n".format(index, port),
> >                       enumerate(portList)
>                     )
>                 )
> E           TypeError: <lambda>() takes exactly 2 arguments (1 given)
> 
> duthost    = <MultiAsicSonicHost arc-switch1038>
> duthosts   = [<MultiAsicSonicHost arc-switch1038>]
> file       = <closed file '/tmp/default_interface_to_front_map.ini', mode 'w' at 0x7f60d028cd20>
> intfInfo   = {'Ethernet0': {'admin_state': u'up', 'alias': u'etp1', 'fec': u'rs', 'name': u'Ethernet0', ...}, 'Ethernet100': {'admi...et104', ...}, 'Ethernet108': {'admin_state': u'up', 'alias': u'etp28', 'fec': u'rs', 'name': u'Ethernet108', ...}, ...}
> port       = 'Ethernet68'
> portList   = ['Ethernet0', 'Ethernet4', 'Ethernet8', 'Ethernet12', 'Ethernet16', 'Ethernet20', ...]
> portMapFile = '/tmp/default_interface_to_front_map.ini'
> ptfhost    = <tests.common.devices.ptf.PTFHost object at 0x7f60d0924590>
> rand_one_dut_hostname = 'arc-switch1038'
> 
> common/fixtures/ptfhost_utils.py:186: TypeError
> >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> 
> >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> > /root/mars/workspace/sonic-mgmt/tests/common/fixtures/ptfhost_utils.py(186)ptf_portmap_file()
> -> enumerate(portList)
> (Pdb) portList
> [u'Ethernet0', u'Ethernet4', u'Ethernet8', u'Ethernet12', u'Ethernet16', u'Ethernet20', u'Ethernet24', u'Ethernet28', u'Ethernet32', u'Ethernet36', u'Ethernet40', u'Ethernet44', u'Ethernet48', u'Ethernet52', u'Ethernet56', u'Ethernet60', u'Ethernet64', u'Ethernet65', u'Ethernet66', u'Ethernet67', u'Ethernet68', u'Ethernet72', u'Ethernet76', u'Ethernet80', u'Ethernet82', u'Ethernet84', u'Ethernet88', u'Ethernet92', u'Ethernet96', u'Ethernet100', u'Ethernet104', u'Ethernet108', u'Ethernet112', u'Ethernet116', u'Ethernet120', u'Ethernet124']
> (Pdb) index
> *** NameError: name 'index' is not defined
> (Pdb) port
> u'Ethernet68'
> ```

Hi @ppikh. #5605 should fix this issue.

**Pull request merged**